refactor vault configuration and add master-key KMS

aead opened the pull request:
## Description
This refactors the vault configuration by moving the
vault-related environment variables to `environment.go`
(Other ENV should follow in the future to have a central
place for adding / handling ENV instead of magic constants
and handling across different files)

Further this commit adds master-key SSE-S3 support.
The operator can specify a SSE-S3 master key using
`MINIO_SSE_MASTER_KEY` which will be used as master key
to derive and encrypt per-object keys for SSE-S3
requests.

This commit is also a pre-condition for SSE-S3
auto-encyption support.
## Motivation and Context
SSE-S3, KMS, Fixes #6329 

## Regression
No

## How Has This Been Tested?
manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

poornas commented on docs/kms/README.md:
````diff
@@ -26,7 +28,26 @@ export MINIO_SSE_VAULT_KEY_NAME=my-minio-key
+export MINIO_KMS_MASTER_KEY=MasterKey1:6368616e676520746869732070617373776f726420746f206120736563726574
+``` 
+
+**Important:** If you loose your master key you won't be able to decrypt the objects anymore. This means you lost
````
s/loose/lose

aead replied:
Fixed

poornas commented on docs/kms/README.md:
```diff
@@ -26,7 +28,26 @@ export MINIO_SSE_VAULT_KEY_NAME=my-minio-key
-### 4. Test your setup
+#### 2.2 Direct master key
+
+Alternatively you can specify a the master key directly using:
```
typo: s/a the/a

aead replied:
Fixed

poornas commented on docs/kms/README.md:
```diff
@@ -26,7 +28,26 @@ export MINIO_SSE_VAULT_KEY_NAME=my-minio-key
+
+**Notice:** The master key is accessible for anyone who has access to the Minio server. In general it is recommended to use
+an external KMS service (Vault). From a security point of view specifying a KMS master key directly is only useful if the storage
+backend is not at the same machine as the Minio server instance - for example Minio in gateway mode.  
```
> for example Minio in gateway mode.  
>
since encryption is disabled for all gateways with the exception of S3 (once PR #6423 is in), we might want to give a different usecase for specifying KMS master key directly to avoid confusion.

aead replied:
Well, that's the primary use case: User runs Minio Gateway pointing to S3 to use AWS as scalable storage provider while keeping object content private. (No trust in AWS required). Other use cases - like storage is on a network drive - is also possible but no clue how common.

poornas replied:
ok

Commit:
refactor vault configuration and add master-key KMS

This refactors the vault configuration by moving the
vault-related environment variables to `environment.go`
(Other ENV should follow in the future to have a central
place for adding / handling ENV instead of magic constants
and handling across different files)

Further this commit adds master-key SSE-S3 support.
The operator can specify a SSE-S3 master key using
`MINIO_SSE_MASTER_KEY` which will be used as master key
to derive and encrypt per-object keys for SSE-S3
requests.

This commit is also a pre-condition for SSE-S3
auto-encyption support.

Fixes #6329

poornas commented on cmd/config-current.go:
```diff
@@ -473,12 +473,8 @@ func (s *serverConfig) loadToCachedConfigs() {
-			globalKMSKeyID = globalKMSConfig.Vault.Key.Name
-		}
+	if err := Environment.LookupKMSConfig(s.KMS); err != nil {
+		logger.FatalIf(err, "Unable to setup the KMS")
```
This lookup should be done from config.json only if globalKMS is not initialized.Otherwise, it will give ambiguous config error because globalKMS is already initialized.

aead replied:
What is the scope of `loadToCachedConfigs()`?
From my understanding it should set the global configuration based on a cached `serverConfig` struct. Shouldn't it overwrite global config in any case, not just if it isn't set?

poornas replied:
When env variables for KMS have been set, globalKMS has already been initialized in loadFromEnvs. When a config entry for KMS exists, this fails with "Ambiguous KMS configuration: vault configuration and a master key are provided at the same time" because globalKMS is already initialized. So,this check needs to be done only if globalKMS is not initialized,since config from env takes priority over config.json entry

aead replied:
When ENV vars are set, `LookupKMSConfig(...)` doesn't honor config file values. It looks up the ENV var, if present it parses and uses the value, otherwise it uses the conig file value.
`Ambiguous KMS configuration: vault configuration and a master key are provided at the same time` should only occur if the `MINIO_SSE_MASTER_KEY` **AND** an non-empty vault config is present (either through ENV or config file).

If `globalKMS != nil ` check is added the globalKMS would never be changed if config is changed and reloaded.

poornas replied:
@aead, this would work if LookupKMSConfig(...) is done once. To repro this issue,  start with vault env settings and no ```MINIO_SSE_MASTER_KEY``` and start minio from a fresh directory. I get a ```"Ambiguous KMS config.."``` error because the config migration creates an empty kms config in loadFromEnvs, and globalKMS gets initialized with vault config from envs. The loadToCachedConfigs is executed after that, and fails because globalKMS is not nil, and so is the server config.Here, you might want to load from cached config entries only if the globalKMS has not been initialized through env vars.

aead replied:
@poornas 1. Can you retry the steps after fetching the latest version of the commit - I've fixed a bug in `LookupKMSConfig(...)`? So you set vault env and start the server. Than you kill it and restart it using a different directory (causing config creation / upgrade)? With the current version I don't see this issue...

My point here is that loading the server configuration should **not** depend on the current state of the server. It should only honor all configuration options and pass if they are valid. That leads to a *stateless* `LookupKMSConfig(...)` function. With `globalKMS != nil` check you skip a config lookup which may should have failed b/c of incorrect configuration.

poornas replied:
it works now with the latest commit @aead.

poornas reviewed (dismissed):
Tested, LGTM.

harshavardhana reviewed (approved):
LGTM, neat idea with Environment style @aead

harshavardhana reviewed (changes requested):
I see that doc was removed, can we add documentation regarding using this feature @aead ?

poornas commented on docs/kms/README.md:
````diff
@@ -64,16 +89,24 @@ export MINIO_SSE_VAULT_KEY_NAME=my-minio-key
-Optionally set `VAULT_NAMESPACE` if AppRole and Transit Secrets engine have been scoped to Vault Namespace
-```
-export VAULT_NAMESPACE=ns1
+Note: If [Vault Namespaces](https://learn.hashicorp.com/vault/operations/namespaces) are in use, VAULT_NAMESPACE variable needs to be set before setting approle and transit secrets engine.
````
This needs to be renamed to MINIO_SSE_VAULT_NAMESPACE

poornas reviewed (approved):
LGTM

harshavardhana committed:
Merge branch 'master' into master-key-kms

harshavardhana committed:
Merge branch 'master' into master-key-kms
